### PR TITLE
sql: consolidate vars Option implementations

### PR DESCRIPTION
**Pull request by guswynn**
Cleanup from https://github.com/MaterializeInc/materialize/pull/19176

### Motivation

   * This PR refactors existing code.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

### CODE REVIEW

File: src/sql/src/session/vars.rs [hunk 1]
```diff
@@ -2871,52 +2871,16 @@ impl Value for Vec<Ident> {
     }
 }
 
-impl Value for Option<String> {
+// Implement `Value` for `Option<V>` for any owned `V`.
+impl<V: Value + Clone + ToOwned<Owned = V>> Value for Option<V> {
```
Comment (benesch) [inline src/sql/src/session/vars.rs:2875]: This isn't officially part of our style guide (not yet, at least) but I find it easier to read complicated bounds with a dedicated `where` clause:

```suggestion
impl<V> Value for Option<V>
where
  V: Value + Clone + ToOwned<Owned = V>
{
```

File: src/sql/src/session/vars.rs [hunk 2]
```diff
@@ -2871,52 +2871,16 @@ impl Value for Vec<Ident> {
...
+    // `Option<V>` value parameters require string LaunchDarkly types.
     const TYPE_NAME: &'static str = "optional string";
```
Comment (benesch) [inline src/sql/src/session/vars.rs:2877]: This seems funny! The type name in Materialize should have as much precision as possible, even if we have to cheat a bit in LD.

```suggestion
    const TYPE_NAME: &'static str = concat!("optional ", V::TYPE_NAME);
```

**Review by benesch — APPROVED** [review]
Very nice! Thanks for doing the cleanup.

**Review by aalexandrov — APPROVED** [review]
:+1: for Nikhil's suggestions.

File: src/sql/src/session/vars.rs (hunk 2 above)
Comment (aalexandrov) [reply src/sql/src/session/vars.rs:2877]: Note that the `TYPE_NAME` also shows up for non-private `Option<T>` session / server vars when you type `SHOW ALL` as a user.

File: src/sql/src/session/vars.rs (hunk 1 above)
Comment (guswynn) [reply src/sql/src/session/vars.rs:2875]: yeah this looks better! We should update the style guide...

File: src/sql/src/session/vars.rs (hunk 2 above)
Comment (guswynn) [reply src/sql/src/session/vars.rs:2877]: Oh i forgot about `concat!`!!

Comment (guswynn) [reply src/sql/src/session/vars.rs:2877]:
```
error[E0401]: can't use generic parameters from outer function
    --> src/sql/src/session/vars.rs:2909:67
     |
2905 | impl<V> Value for Option<V>
     |      - type parameter from outer function
...
2909 |     const TYPE_NAME: &'static str = concatcp!("optional string", <V as Value>::TYPE_NAME);
     |                                                                   ^ use of generic parameter from outer function

note: erroneous constant used
    --> src/sql/src/session/vars.rs:2909:37
     |
2909 |     const TYPE_NAME: &'static str = concatcp!("optional string", <V as Value>::TYPE_NAME);
     |                                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
     |
     = note: this note originates in the macro `$crate::pmr::__concatcp_impl` which comes from the expansion of the macro `concatcp` (in Nightly builds, run with -Z macro-backtrace for more info)

note: erroneous constant used
    --> src/sql/src/session/vars.rs:2909:37
     |
2909 |     const TYPE_NAME: &'static str = concatcp!("optional string", <V as Value>::TYPE_NAME);
     |                                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
     |
     = note: this note originates in the macro `__cf_osRcTFl4A::__concatcp_inner` which comes from the expansion of the macro `concatcp` (in Nightly builds, run with -Z macro-backtrace for more info)

For more information about this error, try `rustc --explain E0401`.
error: could not compile `mz-sql` due to previous error
warning: build failed, waiting for other jobs to finish...
    Building [=======================> ] 771/776: clusterd(bin)
    
  
```

its unclear to me if this possible at all in rust right now :(

File: src/sql/src/session/vars.rs (hunk 1 above)
Comment (benesch) [reply src/sql/src/session/vars.rs:2927]: (This looks unresolved!)

**Review by benesch — APPROVED** [review]
Beaut, thanks!

File: src/sql/src/session/vars.rs (hunk 1 above)
Comment (guswynn) [reply src/sql/src/session/vars.rs:2927]: lost in my `undo` spree

**Pull request merged**